text/0004-configurational-dependencies.md

```diff
@@ -0,0 +1,57 @@
+# Configurational dependencies
+
+## Summary
+
+A new type of dependencies for installing configuration and hooks.
+
+## Motivation
+
+We want to make it possible to share some configurations between projects. Storing the configuration in regular dependencies is not an option as we might need the configuration during installation of the "regular" dependencies ("dependencies", "devDependencies", and "optionalDependencies"). Hence, we need to install these configurational dependencies before other types of dependencies.
+
+Some examples of usage:
+
+* Installing hooks used by [`.pnpmfile.cjs`](https://pnpm.io/pnpmfile).
+* Installing the list of dependencies that are allowed to be built (the [pnpm.onlyBuiltDependenciesFile](https://pnpm.io/package_json#pnpmonlybuiltdependenciesfile)).
+* Installing [catalogs](https://pnpm.io/catalogs).
+* Loading patch file.
+
+## Detailed Explanation
+
+There will be a new field in `package.json` called `pnpm.configDependencies`. For example:
+
+```json
+{
+  "name": "my-pkg",
+  "version": "0.0.0",
+  "pnpm": {
+    "configDependencies": {
+      "my-configs": "1.0.0+sha512-30iZtAPgz+LTIYoeivqYo853f02jBYSd5uGnGpkFV0M3xOt9aN73erkgYAmZU43x4VfqcnLxW9Kpg3R5LC4YYw=="
+    }
+  }
+}
+```
+
+These new type of "configurational" dependencies will be npm packages with a lot of limitations:
+
+* They won't have any dependencies. Even if they will have dependencies, pnpm will ignore them during installation.
+* They will not have lifecycle scripts.
+* They will only be installable via exact versions.
+
+These dependencies will be installed into a new directory (name to be decided), not into `node_modules`.
+
+## Rationale and Alternatives
+
+For the list of "onlyBuiltDependencies" currently we can load them from `node_modules`. However, that does introduce some "works on my machine" issues as we don't have list during the linking from store stage. As a result, if the dependency is not allowed to be built, we still might load it from the side-effects cache.
```

Comment:
nitpick: I think there's a word missing in one of the sentences here.

```suggestion
For the list of "onlyBuiltDependencies" currently we can load them from `node_modules`. However, that does introduce some "works on my machine" issues as we don't have the list during the linking from store stage. As a result, if the dependency is not allowed to be built, we still might load it from the side-effects cache.
```

I would tweak the wording of the paragraph a bit more, but feel free to ignore this suggestion if you disagree:

```suggestion
The `"onlyBuiltDependencies"` list can currently be loaded from `node_modules`. However, this introduces "works on my machine" issues since the list isn't available during the link-from-store stage. As a result, if the dependency is not allowed to be built, it might still be loaded from the side-effects cache.
```